### PR TITLE
Day 2

### PR DESCRIPTION
**Pull request by pviafore**

### CODE REVIEW

File: common/input_file.py [hunk 1]
```diff
@@ -7,5 +7,18 @@ def read_numbers(input_filename):
...
+    return get_transformed_input(input_filename, str)
+
+def get_transformed_input(input_filename, transform_function):
+    """
+        Read all the lines of a file and transform it according to
+        transform_function
+    """
     with open(input_filename) as input_file:
-        return [int(n) for n in input_file.readlines()]
+        return [transform_function(line).strip() for line in input_file.readlines()]
```
Comment (jmhossler) [inline common/input_file.py]: read_numbers should fail now. You should call strip() on line, not the result of the transform, OR you could pass in a lambda in read_strings to do the strip after calling str.

File: challenge2.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        assert False, "We did not find a suitable answer in the input"
+
+def is_off_by_one_letter(candidate):
+    """
+        Return true if the boxes are off by one letter
+    """
+    box1, _, common_letters = candidate
+    return len(common_letters) == len(box1) - 1
+
+def get_common_letters(box1, box2):
```
Comment (jmhossler) [inline challenge2.py]: I like how you have the logic for getting common letters separate from getting the correct boxes.

File: challenge2.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+def get_correct_boxes(box_ids):
+    """
+        Get the pair of correct boxes
+    """
+    box_pairs = combinations(box_ids, 2)
+
+    # get the common letters for each box so that we can see which ones are only one off
+    candidates = [(box1, box2, get_common_letters(box1, box2)) for box1, box2 in box_pairs]
+    try:
+        return next(candidate for candidate in candidates if is_off_by_one_letter(candidate))
```
Comment (jmhossler) [inline challenge2.py]: it feels weird to return the common letters of the correct boxes. I know it's convenient based on the implementation, but I would be confused with all the output this function returns. I would expect it to just return the correct boxes based on the name.

File: challenge2.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    """
+        Gets the count of boxes where we have exactly `desired_count`
+        number of one letter in it
+    """
+    candidates = [True for box in grouped_box_ids if contains_desired_count(box, desired_count)]
+    return len(candidates)
+
+def contains_desired_count(box, desired_count):
+    """
+        If the box has exatly 'desired_count` of any one letter
```
Comment (jmhossler) [inline challenge2.py]: one issue with comments is that it allows inconsistent spelling mistakes ;)

File: challenge2.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+def transform_input(box_id):
+    """
+        We sort and groupby to get lists of letters from each ID
+    """
+    grouped_box_id = groupby(sorted(box_id))
+
+    # We need to convert to a list so that we may go through this
+    # list more than once, otherwise we exhaust iterators first time through
+
+    return [(key, list(group)) for key, group in grouped_box_id]
```
Comment (jmhossler) [inline challenge2.py]: since all you need is the frequency here, could we just go ahead and use len(group)?

File: challenge2.py (hunk 4 above)
Comment (pviafore) [reply challenge2.py]: yup, thanks

File: challenge2.py (hunk 3 above)
Comment (pviafore) [reply challenge2.py]: yeah, that's a good point.  I think the function is trying to do too much

File: challenge2.py (hunk 2 above)
Comment (pviafore) [reply challenge2.py]: thanks

File: common/input_file.py (hunk 1 above)
Comment (pviafore) [reply common/input_file.py]: I think I always want the \n stripped off

File: challenge2.py (hunk 5 above)
Comment (pviafore) [reply challenge2.py]: That's a good improvement, thanks!

I'm going to keep it as is to reflect my original algorithm, but I think you're absolutely right.

**Pull request merged**